dns/dns_updater.go

```diff
@@ -0,0 +1,97 @@
+package dns
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/sky-uk/feed/controller"
+	"github.com/sky-uk/feed/dns/r53"
+	"github.com/sky-uk/feed/elb"
+	"k8s.io/heapster/Godeps/_workspace/src/github.com/coreos/fleet/log"
+)
+
+type updater struct {
+	dns          r53.Route53Client
+	elb          elb.ELB
+	elbLabelName string
+}
+
+// New creates an updater for dns
+func New(dnsClient r53.Route53Client, awsElb elb.ELB, elbLabelName string) controller.Updater {
+	return &updater{
+		dns:          dnsClient,
+		elb:          awsElb,
+		elbLabelName: elbLabelName,
+	}
+}
+
+func (u *updater) Start() error {
+
+	return nil
+}
+
+func (u *updater) Stop() error {
+
+	return nil
+}
+
+func (u *updater) Health() error {
+
+	return nil
+}
+
+func (u *updater) Update(update controller.IngressUpdate) error {
+	frontEnds, err := elb.FindFrontEndElbs(u.elb, u.elbLabelName)
+	if err != nil {
+		return fmt.Errorf("unable to find front end load balancers: %v", err)
+	}
+
+	aRecords := u.dns.GetARecords()
+
+	changes, err := calculateChanges(frontEnds, aRecords, update)
+	if err != nil {
+		return err
+	}
+
+	u.dns.UpdateRecordSets(changes)
+
+	return nil
+}
+
+func calculateChanges(frontEnds map[string]elb.LoadBalancerDetails, aRecords []*route53.ResourceRecordSet, update controller.IngressUpdate) ([]*route53.Change, error) {
+	changes := []*route53.Change{}
+	hostToIngresEntry := make(map[string]controller.IngressEntry)
+	for _, ingressEntry := range update.Entries {
+		hostToIngresEntry[ingressEntry.Host] = ingressEntry
+		frontEnd, exists := frontEnds[ingressEntry.FrontendName]
+		if !exists {
+			return nil, fmt.Errorf("unable to find front end loadd balancer with name: %v", ingressEntry.FrontendName)
+		}
+		changes = append(changes, newChange("UPSERT", ingressEntry.Host, frontEnd.DNSName, frontEnd.HostedZoneID))
+	}
+
+	for _, recordSet := range aRecords {
+		if _, contains := hostToIngresEntry[*recordSet.Name]; !contains {
+			changes = append(changes, newChange("DELETE", *recordSet.Name, *recordSet.AliasTarget.DNSName, *recordSet.AliasTarget.HostedZoneId))
+		}
+	}
+
+	log.Infof("calculated changes to dns: %v", changes)
```

Comment:
I am finding the full output very useful during development / debugging. I will remove this in a week :)